pkg/coscheduling/coscheduling.go

```diff
@@ -0,0 +1,279 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package coscheduling
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"strconv"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/api/v1/pod"
+	framework "k8s.io/kubernetes/pkg/scheduler/framework/v1alpha1"
+)
+
+// Coscheduling is a plugin that implements the mechanism of gang scheduling.
+type Coscheduling struct {
+	FrameworkHandle framework.FrameworkHandle
+	PodLister       corelisters.PodLister
+	PodGroupInfos   map[string]PodGroupInfo // Key is the name of Namespace/PodGroup.
+	AutoID          uint64
+}
+
+type PodGroupInfo struct {
+	// InitialTimestamp stores the timestamp of the initialization time of PodGroup.
+	InitialTimestamp time.Time
+	ID               uint64
+	MinAvailable     int
```

Comment:
MinAvailable and Name fields can be removed. The former is not used anywhere, the latter can be deduced from key easily.